frontend/tests/playwright/msw/request-handler.ts

```diff
@@ -0,0 +1,233 @@
+import { isRecord } from '@ifixit/helpers';
+import {
+   GraphQLHandler,
+   RestHandler,
+   RequestHandler,
+   DefaultBodyType,
+   GraphQLContext,
+   GraphQLRequest,
+   GraphQLVariables,
+   PathParams,
+   ResponseResolver,
+   RestContext,
+   RestRequest,
+} from 'msw';
+
+/**
+ * These methods are a subset of the GraphQL operational methods that are
+ * supported by MSW's `graphql` handler. There are of course other methods
+ * `MSW` supports, but we are only interested in these two for now.
+ *
+ * @see https://mswjs.io/docs/api/graphql#operation-kind
+ */
+export type GraphQLMethods = 'mutation' | 'query';
+/**
+ * This is the type of the resolver that is used by the `graphql` handler.
+ * The resolver is a function that accepts a captured request and may return
+ * a mocked response.
+ * @see https://mswjs.io/docs/basics/response-resolver
+ */
+export type GraphQLResolver = ResponseResolver<
+   GraphQLRequest<GraphQLVariables>,
+   GraphQLContext<Record<string, any>>
+>;
+/**
+ * These methods are a subset of the REST methods that are
+ * supported by MSW's `rest` handler - _with the exception of
+ * `options`_.
+ *
+ * @see https://mswjs.io/docs/api/rest#methods
+ */
+export type RestMethods = 'all' | 'get' | 'post' | 'put' | 'patch' | 'delete';
+/**
+ * This is the type of the resolver that is used by the `rest` handler.
+ * The resolver is a function that accepts a captured request and may return
+ * a mocked response.
+ * @see https://mswjs.io/docs/basics/response-resolver
+ */
+export type RestResolver = ResponseResolver<
+   RestRequest<DefaultBodyType, PathParams<string>>,
+   RestContext,
+   DefaultBodyType
+>;
+
+export type RequestInfo = {
+   endpoint: string;
+   method: GraphQLMethods | RestMethods;
+};
+
+/**
+ * This type represents the options that can be passed to the handler
+ * to configure what the response should return if the request is matched.
+ *
+ * @property `status` - The HTTP status code to return
+ * @property `headers` - The headers to return with the response
+ * @property `body` - The body of the response.
+ * @property `responseType` - This the kind of transformation to apply to the
+ * body when it will be returned. This is only applicable to REST requests
+ * since GraphQL responses will only use a `data` transformer.
+ *
+ * @see https://mswjs.io/docs/basics/response-transformer#standards-transformers
+ * @see https://mswjs.io/docs/api/context
+ * @see https://mswjs.io/docs/api/response
+ *
+ */
+export type MockedResponseInfo = {
+   status: number;
+   headers?: Record<string, string | string[]>;
+   body?: Record<string, any> | string | number | boolean | null;
+   responseType?: 'json' | 'text' | 'xml' | 'raw';
+};
```

Comment:
I'd suggest splitting this into two types: one for GraphQL requests and one for REST requests. I think trying to keep them together and then filter the appropriate keys out below is probably not simplifying things. If you want to avoid duplication on `status`, you can always create a `type BaseResponseInfo = { status: number }` and then intersect it into the field set for each subtype.